File: mycollab-web/src/main/scala/com/mycollab/module/project/view/ProjectModuleController.scala
/**
 * This file is part of mycollab-web.
 *
 * mycollab-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-web.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.view

import java.util.GregorianCalendar

import com.mycollab.eventmanager.ApplicationEventListener
import com.mycollab.module.project.events.ProjectEvent.{GotoMyProject, GotoUserDashboard}
import com.mycollab.module.project.events.{ClientEvent, ProjectEvent, ReportEvent, StandUpEvent}
import com.mycollab.module.project.view.client.IClientPresenter
import com.mycollab.module.project.view.parameters.ClientScreenData.{Add, Read}
import com.mycollab.module.project.view.parameters.{ClientScreenData, ProjectScreenData, ReportScreenData, StandupScreenData}
import com.mycollab.module.project.view.reports.IReportPresenter
import com.mycollab.vaadin.mvp.{AbstractController, PageActionChain, PresenterResolver, ScreenData}
import com.google.common.eventbus.Subscribe
import com.mycollab.module.crm.domain.SimpleAccount
import com.mycollab.module.crm.domain.criteria.AccountSearchCriteria

/**
  * @author MyCollab Ltd.
  * @since 5.0.9
  */
class ProjectModuleController(val container: ProjectModule) extends AbstractController {
  this.register(new ApplicationEventListener[ProjectEvent.GotoUserDashboard]() {
    @Subscribe override def handle(event: GotoUserDashboard): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[UserDashboardPresenter])
      presenter.go(container, null)
    }
  })

  this.register(new ApplicationEventListener[ProjectEvent.GotoMyProject]() {
    @Subscribe override def handle(event: GotoMyProject): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[ProjectViewPresenter])
      presenter.handleChain(container, event.getData.asInstanceOf[PageActionChain])
    }
  })

  this.register(new ApplicationEventListener[ProjectEvent.GotoList]() {
    @Subscribe override def handle(event: ProjectEvent.GotoList): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[UserDashboardPresenter])
      presenter.go(container, new ProjectScreenData.GotoList())
    }
  })

  this.register(new ApplicationEventListener[ClientEvent.GotoList]() {
    @Subscribe override def handle(event: ClientEvent.GotoList): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IClientPresenter])
      val searchCriteria = new AccountSearchCriteria
      presenter.go(container, new ClientScreenData.Search(searchCriteria))
    }
  })

  this.register(new ApplicationEventListener[ClientEvent.GotoAdd]() {
    @Subscribe def handle(event: ClientEvent.GotoAdd): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IClientPresenter])
      val account = new SimpleAccount
      presenter.go(container, new Add(account))
    }
  })

  this.register(new ApplicationEventListener[ClientEvent.GotoEdit]() {
    @Subscribe def handle(event: ClientEvent.GotoEdit): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IClientPresenter])
      presenter.go(container, new ScreenData.Edit[Any](event.getData))
    }
  })

  this.register(new ApplicationEventListener[ClientEvent.GotoRead]() {
    @Subscribe def handle(event: ClientEvent.GotoRead): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IClientPresenter])
      val clientId = event.getData.asInstanceOf[Integer]
      presenter.go(container, new Read(clientId))
    }
  })

  this.register(new ApplicationEventListener[ReportEvent.GotoConsole]() {
    @Subscribe override def handle(event: ReportEvent.GotoConsole): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IReportPresenter])
      presenter.go(container, null)
    }
  })

  this.register(new ApplicationEventListener[ReportEvent.GotoTimesheetReport]() {
    @Subscribe override def handle(event: ReportEvent.GotoTimesheetReport): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IReportPresenter])
      presenter.go(container, new ReportScreenData.GotoTimesheet())
    }
  })

  this.register(new ApplicationEventListener[ReportEvent.GotoWeeklyTimingReport]() {
    @Subscribe override def handle(event: ReportEvent.GotoWeeklyTimingReport): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IReportPresenter])
      presenter.go(container, new ReportScreenData.GotoWeeklyTiming())
    }
  })

  this.register(new ApplicationEventListener[ReportEvent.GotoUserWorkloadReport]() {
    @Subscribe override def handle(event: ReportEvent.GotoUserWorkloadReport): Unit = {
      val presenter = PresenterResolver.getPresenter(classOf[IReportPresenter])
      presenter.go(container, new ReportScreenData.GotoUserWorkload())
    }
  })

  this.register(new ApplicationEventListener[StandUpEvent.GotoList] {
    @Subscribe def handle(event: StandUpEvent.GotoList) {
      val presenter = PresenterResolver.getPresenter(classOf[IReportPresenter])
      presenter.go(container, new StandupScreenData.Search(new GregorianCalendar().getTime))
    }
  })
}

File: mycollab-web/src/main/java/com/mycollab/module/project/view/ProjectModuleController.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.view

import com.google.common.eventbus.Subscribe
import com.mycollab.common.domain.SimpleClient
import com.mycollab.common.service.ClientService
import com.mycollab.module.project.event.*
import com.mycollab.module.project.view.parameters.ClientScreenData
import com.mycollab.module.project.view.parameters.ProjectScreenData
import com.mycollab.module.project.view.parameters.ReportScreenData
import com.mycollab.module.project.view.parameters.StandupScreenData
import com.mycollab.spring.AppContextUtil
import com.mycollab.vaadin.AppUI
import com.mycollab.vaadin.ApplicationEventListener
import com.mycollab.vaadin.mvp.AbstractController
import com.mycollab.vaadin.mvp.PageActionChain
import com.mycollab.vaadin.mvp.PresenterResolver
import java.time.LocalDate

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class ProjectModuleController(val container: ProjectModule) : AbstractController() {
    init {
        this.register(object : ApplicationEventListener<ProjectEvent.GotoMyProject> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoMyProject) {
                val presenter = PresenterResolver.getPresenter(ProjectViewPresenter::class.java)
                presenter.handleChain(container, event.data as PageActionChain)
            }
        })

        this.register(object : ApplicationEventListener<ProjectEvent.GotoUserDashboard> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoUserDashboard) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, null)
            }
        })

        this.register(object : ApplicationEventListener<ProjectEvent.GotoList> {
            @Subscribe
            override fun handle(event: ProjectEvent.GotoList) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ProjectScreenData.GotoList())
            }
        })

        this.register(object : ApplicationEventListener<FavoriteEvent.GotoList> {
            @Subscribe
            override fun handle(event: FavoriteEvent.GotoList) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ProjectScreenData.GotoList())
            }
        })

        this.register(object : ApplicationEventListener<ClientEvent.GotoList> {
            @Subscribe
            override fun handle(event: ClientEvent.GotoList) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ClientScreenData.Search(null))
            }
        })

        this.register(object : ApplicationEventListener<ClientEvent.GotoAdd> {
            @Subscribe
            override fun handle(event: ClientEvent.GotoAdd) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                val client = SimpleClient()
                presenter.go(container, ClientScreenData.Add(client))
            }
        })

        this.register(object : ApplicationEventListener<ClientEvent.GotoEdit> {
            @Subscribe
            override fun handle(event: ClientEvent.GotoEdit) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                if (event.data is Int) {
                    val clientService = AppContextUtil.getSpringBean(ClientService::class.java)
                    val client = clientService.findById(event.data, AppUI.accountId)
                    if (client != null) {
                        presenter.go(container, ClientScreenData.Edit(client))
                    }
                } else if (event.data is SimpleClient) {
                    presenter.go(container, ClientScreenData.Edit(event.data))
                }
            }
        })

        this.register(object : ApplicationEventListener<ClientEvent.GotoRead> {
            @Subscribe
            override fun handle(event: ClientEvent.GotoRead) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                val clientId = event.data as Int
                presenter.go(container, ClientScreenData.Read(clientId))
            }
        })

        this.register(object : ApplicationEventListener<ReportEvent.GotoConsole> {
            @Subscribe
            override fun handle(event: ReportEvent.GotoConsole) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ReportScreenData.GotoConsole())
            }
        })

        this.register(object : ApplicationEventListener<ReportEvent.GotoTimesheetReport> {
            @Subscribe
            override fun handle(event: ReportEvent.GotoTimesheetReport) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ReportScreenData.GotoTimesheet())
            }
        })

        this.register(object : ApplicationEventListener<ReportEvent.GotoWeeklyTimingReport> {
            @Subscribe
            override fun handle(event: ReportEvent.GotoWeeklyTimingReport) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ReportScreenData.GotoWeeklyTiming())
            }
        })

        this.register(object : ApplicationEventListener<ReportEvent.GotoUserWorkloadReport> {
            @Subscribe
            override fun handle(event: ReportEvent.GotoUserWorkloadReport) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, ReportScreenData.GotoUserWorkload())
            }
        })

        this.register(object : ApplicationEventListener<StandUpEvent.GotoList> {
            @Subscribe
            override fun handle(event: StandUpEvent.GotoList) {
                val presenter = PresenterResolver.getPresenter(BoardContainerPresenter::class.java)
                presenter.go(container, StandupScreenData.Search(LocalDate.now()))
            }
        })
    }
}